test/integration/http2_upstream_integration_test.cc

```diff
@@ -306,17 +306,24 @@ TEST_P(Http2UpstreamIntegrationTest, UpstreamConnectionCloseWithManyStreams) {
                                                             {":authority", "host"}});
     encoders.push_back(&encoder_decoder.first);
     responses.push_back(std::move(encoder_decoder.second));
+
+    // Ensure we've established a connection before issuing the resets.
+    // TODO(snowp): Why can't we handle resets here?
```

Comment:
Yeah I'm a little confused here. What was the sequence causing the hang here? I don't think we should need to wait for upstream before sending client resets? What was actually hanging?

Reply:
Without this change, the code was hanging in `waitForNewStream`, which we never hit since we reset the stream while it's in the pending queue.

If only call `waitForNewStream` for streams that won't get reset, this ends up hanging in `waitForEndStream` about 50% of the time. The test sends between 1 and 51 requests, reseting every 15th, and every time it fails it's when a reset happens in the middle of all the requests (so number of requests > 15), but it doesn't always fail in that case. 

It seems like it's getting stuck waiting for the (15 + 1)th stream to end, so I was thinking it might be a race condition either in the conn pool or in how we account for stream/connection events in test. It seems like somehow the reset on stream i=15 ends up affecting stream i=16.

I suspected that the pending request queue might be concurrently accessed, but guarding it with a mutex didn't seem to help.

I agree that this should work, but at this point I'm not even sure if the connection pool is buggy or if the tests just need to be reworked, so open to any ideas :)

Reply:
Thinking about it some more I think it might be due to a race between when we call reset and when the connection is established: any reset that comes *after* connect will be proceeded by a new stream (because we've already called onPoolReady), while any reset *before* connect will not trigger a new stream (we'll cancel the pending request).

Reply:
I'm not comfortable merging this change until we understand this. I can think about this more tomorrow morning. Do you have full trace logs for the hang? That would be useful to look at.

Reply:
Here's a gist with a good and bad test run: 
failed run: https://gist.githubusercontent.com/snowp/ecb62fd9cf74dc885f68ddfe2aa0ac8e/raw/8e54586118f15cbcf9b0eb3eb8f1030447eda8e9/failure.out
successful run: https://gist.githubusercontent.com/snowp/ecb62fd9cf74dc885f68ddfe2aa0ac8e/raw/8e54586118f15cbcf9b0eb3eb8f1030447eda8e9/success.out

This is when running with the following modifications to the code in the PR: https://github.com/snowp/envoy/commit/4b590995a0199010575dbd64e86192d5b8b6f3c8

Namely: 
 - move waiting for the connection back to the original location
 - hard code the number of requests instead of picking a random number
 - skip waiting for requests that should be reset (without this we fail on the first waitForNewStream)

The test sends 17 requests, and resets every 15th of them, namely it will reset the 0th and 15th request.

This bit from the output logs makes me think my previous theory about this being a race between when the connection is established and when the request is reset:
```
~/Development/envoy(h2-queue):0 grep cancel failure
[2018-11-02 04:01:20.395][2075503][debug][router] [source/common/router/router.cc:938] [C1][S2898179549770959211] cancelling pool request
[2018-11-02 04:01:20.395][2075503][debug][pool] [source/common/http/conn_pool_base.cc:41] cancelling pending request
~/Development/envoy(h2-queue):0 grep cancel success
[2018-11-02 04:00:15.278][2069333][debug][router] [source/common/router/router.cc:938] [C1][S4194523254301487947] cancelling pool request
[2018-11-02 04:00:15.278][2069333][debug][pool] [source/common/http/conn_pool_base.cc:41] cancelling pending request
[2018-11-02 04:00:15.308][2069333][debug][router] [source/common/router/router.cc:938] [C1][S82244168407272366] cancelling pool request
[2018-11-02 04:00:15.308][2069333][debug][pool] [source/common/http/conn_pool_base.cc:41] cancelling pending request
``` 
Specifically, in the successful run both resets happened **before** the connection completed, as indicated by the log line stating that the pool request was canceled. 

I suspect that in the run that failed, one of the reset requests created a stream (because it wasn't canceled), so it messed up the stream accounting on our end: we didn't expect the 15th request to generate a stream, so when it did we attributed it to the 16th request. This test is now failing because it's waiting for `endStream` on what we think is the 16th stream, but that's actually the 15th request's stream, which has been reset and won't reach `endStream`.

If I'm right, then in order to have this test send resets before the connection has been established, we need a way to tell whether we reset the request before or after the connect, because depending on that we may or may not see a stream on the `FakeConnection`. One option would be to wait for N streams, compare to total number of requests R and from that infer that the first (R-N) requests must have been reset before the connection was established.

Hopefully this makes sense. Would definitely want someone else to look this over to make sure it seems like a reasonable explanation of what we're seeing.

Reply:
OK thanks for the really detailed investigation here. I agree your theory makes sense.

@alyssawilk probably disagrees with me here, but I would be inclined to make this test more deterministic (remove the random number of requests). I think the randomness makes this more confusing than it probably needs to be. 

Can we figure out a way to make this test a bit easier to understand and more deterministic? @alyssawilk can likely help with this on Monday.

Reply:
I'll accept removing randomity here - I care more about it when it comes to payload sizes.  Failing the nth request when n < [default max stream sizes] is less likely than triggering backup due to larger/smaller payloads

Reply:
Updated the test to always send 20 requests, so we get a reset on the 1st and 15th request.